File: kaantelypeli/src/main/java/kaantelypeli/ui/Game.java
package kaantelypeli.ui;

import javafx.animation.AnimationTimer;
import javafx.application.Application;
import javafx.event.EventHandler;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.Pane;
import javafx.scene.layout.VBox;
import javafx.scene.shape.Rectangle;
import javafx.stage.Stage;
import kaantelypeli.engine.Level;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.function.BiFunction;

import static kaantelypeli.utils.FXUtils.createButton;
import static kaantelypeli.utils.FileOperations.loadLevel;

/**
 * Main Graphical User Interface class.
 */
public class Game extends Application {
    public static final int SCALE = 2;
    private Stage mainStage;
    private Scene mainMenu;
    public static final String FILESERVER =  "http://xylix.fi/levels";



    /**
     * Launch the application.
     * @param args pass command line arguments to follow
     * main() method convention
    */
    public static void main(String[] args) {
        Application.launch();
    }
    
    @Override
    public void start(Stage stage) {
        mainStage = stage;
        // Button cloud = createButton("cloud-levels", event -> cloudMenu(mainStage));
        LevelEditor editor = new LevelEditor(stage, mainMenu);
        Button editorButton = createButton("editor", event -> editor.editorMenu());
        // VBox buttons = new VBox(cloud, editorButton);
        VBox buttons = new VBox(editorButton);

        InputStreamReader levelFolder = new InputStreamReader(Game.class.getResourceAsStream("/levels"));
        BufferedReader br = new BufferedReader(levelFolder);
        br.lines().forEach(line -> buttons.getChildren().add(levelButton(line, stage)));

        mainMenu = new Scene(buttons);
        stage.setScene(mainMenu);
        stage.show();
    }


    /* private void cloudMenu(Stage target) {
        ChoiceDialog<String> choice = selector(FILESERVER, target);
        Optional<String> result = choice.showAndWait();
        result.ifPresent(input ->  target.setScene(levelToScene(loadLevel(FILESERVER + '/' + input));
    }
     */

    private Button levelButton(String file, Stage stage) {
        final String levelName = file.replace(".json", "");
        return createButton(levelName, event -> {
            Level activeLevel = loadLevel(levelName);
            stage.setScene(levelToScene(activeLevel));
        });
    }

    public Scene getNextScene(Level l) {
        final int nextLevel = l.levelIndex + 1;
        return switch(nextLevel) {
            case 666 -> this.mainMenu;
            case 1 -> levelToScene(loadLevel("2"));
            case 2 -> levelToScene(loadLevel("3"));
            case 3 -> levelToScene(loadLevel("4"));
            case 4 -> levelToScene(loadLevel("5"));
            case 5 -> this.mainMenu;
            default -> throw new IllegalStateException("Next level cannot be negative or more than amount of levels. Was %s".formatted(nextLevel));
        };
    }

    /**
     * Abstraction for hotkey binding to enable refactoring it to other callsite later.
     * Result of bindHotkey needs to be passed as an argument to scene.setOnKeyPressed
     * To bind hotkeys to the related level correctly
     */
    private EventHandler<? super KeyEvent> bindHotkey(Level level, Pane pane) {
        return (event) -> {
            final KeyCode pressedKey = event.getCode();
            switch (pressedKey) {
                case LEFT, A -> level.changeGravity(270);
                case RIGHT, D -> level.changeGravity(90);
                case R -> level.restart();
                case ESCAPE -> mainStage.setScene(mainMenu);
            }
            pane.setRotate(level.gravity);
        };
    }

    public Scene levelToScene(Level level) {
        Pane pane = new Pane(level.getHitboxes().toArray(Rectangle[]::new));
        Scene scene = new Scene(pane);

        scene.setOnKeyPressed(bindHotkey(level, pane));
        new AnimationTimer() {
            @Override public void handle(long timestamp) {
                Level.State gameState = level.tick();

                if (gameState == Level.State.WON) {
                    this.stop();
                    mainStage.setScene(getNextScene(level));
                }
            }
        }.start();
        return scene;
    }

}

File: kaantelypeli/src/main/java/kaantelypeli/ui/Game.kt
package kaantelypeli.ui

import javafx.animation.AnimationTimer
import javafx.application.Application
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.Scene
import javafx.scene.control.Button
import javafx.scene.control.ChoiceDialog
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.Pane
import javafx.scene.layout.VBox
import javafx.stage.Stage
import kaantelypeli.engine.Level
import kaantelypeli.utils.loadLevel
import kaantelypeli.utils.createButton
import kaantelypeli.utils.selector
import java.io.BufferedReader
import java.io.InputStreamReader

/**
 * Main Graphical User Interface class.
 */
class Game : Application() {
    private var mainStage: Stage? = null
    private var mainMenu: Scene? = null
    override fun start(stage: Stage) {
        mainStage = stage
        val cloud: Button = createButton("cloud-levels") { event -> cloudMenu() }
        val editor = LevelEditor(stage, mainMenu)
        val editorButton: Button = createButton("editor") { event -> editor.editorMenu() }
        val buttons = VBox(cloud, editorButton)
        val levelFolder =
            InputStreamReader(Game::class.java.getResourceAsStream("/levels"))
        val br = BufferedReader(levelFolder)
        br.lines().forEach { line: String ->
            buttons.children.add(levelButton(line, stage))
        }
        mainMenu = Scene(buttons)
        stage.scene = mainMenu
        stage.show()
    }

    private fun cloudMenu() {
        val choice: ChoiceDialog<String> = selector(FILESERVER, mainStage)
        val result = choice.showAndWait()
        result.ifPresent { input: String ->
            mainStage!!.scene = toScene(
                loadLevel("$FILESERVER/$input")
            )
        }
    }

    private fun levelButton(file: String, stage: Stage): Button {
        val levelName = file.replace(".json", "")
        val button = createButton(levelName) { _ ->
            val activeLevel = loadLevel(levelName)
            stage.scene = toScene(activeLevel)
        }
        return button
    }

    private fun toScene(level: Level): Scene {
        val pane =
            Pane(*level.hitboxes.toTypedArray())
        val scene = Scene(pane)
        scene.onKeyPressed = EventHandler { event: KeyEvent ->
            val pressedKey = event.code
            if (pressedKey == KeyCode.LEFT || pressedKey == KeyCode.A) {
                level.changeGravity(270)
                pane.rotate = pane.rotate + 270
            } else if (pressedKey == KeyCode.RIGHT || pressedKey == KeyCode.D) {
                level.changeGravity(90)
                pane.rotate = pane.rotate + 90
            } else if (pressedKey == KeyCode.R) {
                level.restart()
            }
        }
        object : AnimationTimer() {
            override fun handle(timestamp: Long) {
                level.tick()
                level.ifOver(Runnable { mainStage!!.scene = mainMenu })
            }
        }.start()
        return scene
    }

    companion object {
        const val SCALE = 2
        const val FILESERVER = "http://xylix.fi/levels"

        /**
         * Launch the application.
         * @param args pass command line arguments to follow
         * main() method convention
         */
        @JvmStatic
        fun main(args: Array<String>) {
            launch()
        }
    }
}
